storage/remote/read.go

```diff
@@ -69,12 +76,88 @@ func (q *querier) Select(p *storage.SelectParams, matchers ...*labels.Matcher) (
 	remoteReadGauge.Inc()
 	defer remoteReadGauge.Dec()
 
-	res, err := q.client.Read(q.ctx, query)
+	ctx, cancel := context.WithTimeout(q.ctx, q.client.timeout)
+	defer cancel()
+
+	res, err := q.client.Read(ctx, query)
 	if err != nil {
 		return nil, nil, fmt.Errorf("remote_read: %v", err)
 	}
 
-	return FromQueryResult(res), nil, nil
+	contentType := res.Header.Get("Content-Type")
+	if strings.HasPrefix(contentType, "application/x-protobuf") {
+		return q.handleSampledResponse(res)
+	}
+
+	if !strings.HasPrefix(contentType, "application/x-streamed-protobuf; proto=prometheus.ChunkedReadResponse") {
+		return nil, nil, fmt.Errorf("not supported remote read content type: %s", contentType)
+	}
+	return q.handleStreamedResponse(res)
+}
+
+func (q *querier) handleSampledResponse(httpResp *http.Response) (storage.SeriesSet, storage.Warnings, error) {
+	compressed, err := ioutil.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error reading response: %v", err)
+	}
+
+	uncompressed, err := snappy.Decode(nil, compressed)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error decoding response: %v", err)
+	}
+
+	var resp prompb.ReadResponse
+	err = proto.Unmarshal(uncompressed, &resp)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to unmarshal response body: %v", err)
+	}
+
+	if len(resp.Results) != 1 {
+		return nil, nil, fmt.Errorf("responses: want 1, got %d", len(resp.Results))
+	}
+
+	return FromQueryResult(resp.Results[0]), nil, nil
+}
+
+func (q *querier) handleStreamedResponse(httpResp *http.Response) (storage.SeriesSet, storage.Warnings, error) {
+	var (
+		tmp        []string
+		lastSeries string
+		currSeries string
+	)
+
+	seriesNum := -1
+	seriesSet := &chunkSeriesSet{}
+	stream := NewChunkedReader(httpResp.Body, DefaultChunkedReadLimit, nil)
+	for {
+		res := &prompb.ChunkedReadResponse{}
+		err := stream.NextProto(res)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, nil, fmt.Errorf("get next proto from stream failed: %v", err)
+		}
+
+		// TODO: log if return non 1 series in frame?
+
+		for _, series := range res.ChunkedSeries {
+			tmp := tmp[:0]
+			for _, l := range series.Labels {
+				tmp = append(tmp, l.String())
+			}
+			currSeries = strings.Join(tmp, ";")
+			if currSeries != lastSeries {
+				lastSeries = currSeries
+				seriesNum++
+				seriesSet.set = append(seriesSet.set, &chunkSeries{labels: series.Labels, mint: q.mint, maxt: q.maxt})
+			}
+
+			seriesSet.set[seriesNum].chunks = append(seriesSet.set[seriesNum].chunks, series.Chunks...)
```

Comment:
Do we really need to maintain `seriesNum`? It's as easy as getting last item in the slice, right?